File: go-backend-interview-prep/09_performance_and_profiling/06_gc_and_latency_notes/main.go
package main

import (
	"fmt"
	"runtime"
	"time"
)

func main() {
	fmt.Println("=== GC and Latency Notes ===")
	fmt.Println()

	// Baseline
	runtime.GC()
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	baseGC := m.NumGC
	fmt.Printf("  Baseline: NumGC=%d, HeapAlloc=%d KB\n\n", m.NumGC, m.HeapAlloc/1024)

	// Allocate lots of small objects to trigger GC
	fmt.Println("  Allocating 1M small objects...")
	start := time.Now()
	var sink []*[64]byte
	for i := 0; i < 1_000_000; i++ {
		obj := new([64]byte)
		sink = append(sink, obj)
	}
	allocTime := time.Since(start)

	runtime.ReadMemStats(&m)
	fmt.Printf("  After alloc: NumGC=%d (+%d), HeapAlloc=%d KB\n",
		m.NumGC, m.NumGC-baseGC, m.HeapAlloc/1024)
	fmt.Printf("  Alloc time: %s\n\n", allocTime)

	// Release references and force GC
	sink = nil
	runtime.GC()
	runtime.ReadMemStats(&m)
	fmt.Printf("  After release + GC: HeapAlloc=%d KB\n", m.HeapAlloc/1024)
	fmt.Printf("  Total GC pause: %s\n", time.Duration(m.PauseTotalNs))
	fmt.Printf("  NumGC total: %d\n\n", m.NumGC)

	// Show recent pause times
	fmt.Println("  Recent GC pauses (most recent first):")
	for i := 0; i < 5; i++ {
		idx := (int(m.NumGC) - 1 - i) % 256
		if idx < 0 {
			break
		}
		pause := time.Duration(m.PauseNs[idx])
		fmt.Printf("    GC #%d: %s\n", int(m.NumGC)-i, pause)
	}

	fmt.Println()
	fmt.Println("--- Key GC concepts ---")
	fmt.Println("  Go GC: concurrent mark-and-sweep")
	fmt.Println("    - Tri-color algorithm (white/grey/black)")
	fmt.Println("    - Mostly concurrent -- very short STW pauses")
	fmt.Println("    - STW pauses typically < 1ms")
	fmt.Println()
	fmt.Println("  GOGC (GC trigger):")
	fmt.Println("    GOGC=100 (default): GC when heap = 2x live data")
	fmt.Println("    GOGC=200: GC when heap = 3x live data (fewer GCs, more memory)")
	fmt.Println("    GOGC=off: disable GC (careful!)")
	fmt.Println()
	fmt.Println("  Reduce GC pressure:")
	fmt.Println("    1. Preallocate slices: make([]T, 0, n)")
	fmt.Println("    2. Use value types (not pointers) in hot paths")
	fmt.Println("    3. sync.Pool for frequently allocated/freed objects")
	fmt.Println("    4. strings.Builder instead of += concat")
	fmt.Println("    5. Reduce allocation rate = fewer GC cycles")
}

File: go-backend-interview-prep/09_performance_and_profiling/06_gc_and_latency_notes/main.py
"""GC and latency notes -- Python equivalent of the Go example.

Python uses reference counting + cyclic garbage collector.
"""

import gc
import sys
import time


class Node:
    """Simple object for allocation demos."""
    __slots__ = ("value", "next")

    def __init__(self, value: int):
        self.value = value
        self.next = None


def main() -> None:
    print("=== GC and Latency Notes (Python) ===\n")

    # Show GC generation stats
    gc.collect()
    stats = gc.get_stats()
    for i, gen in enumerate(stats):
        print(f"  Generation {i}: {gen}")
    print()

    # Allocate many objects
    print("  Allocating 500K objects...")
    start = time.perf_counter()
    objects = []
    for i in range(500_000):
        objects.append(Node(i))
    alloc_time = time.perf_counter() - start
    print(f"  Alloc time: {alloc_time:.4f}s")
    print()

    # Create reference cycles (to trigger cyclic GC)
    print("  Creating 10K reference cycles...")
    cycles = []
    for i in range(10_000):
        a = Node(i)
        b = Node(i + 1)
        a.next = b
        b.next = a  # cycle!
        cycles.append(a)
    print()

    # Release references
    del objects
    del cycles

    # Force cyclic GC and time it
    gc.collect()  # clear any pending
    start = time.perf_counter()
    collected = gc.collect()
    gc_time = time.perf_counter() - start
    print(f"  gc.collect(): freed {collected} cyclic objects in {gc_time*1000:.2f}ms")
    print()

    # Show reference counting
    obj = Node(42)
    print(f"  Reference counting demo:")
    print(f"    refcount(obj):          {sys.getrefcount(obj)}")  # 2 (obj + arg)
    ref = obj
    print(f"    After ref=obj:          {sys.getrefcount(obj)}")  # 3
    del ref
    print(f"    After del ref:          {sys.getrefcount(obj)}")  # 2
    print(f"    When refcount -> 0: immediately freed (no GC needed)")
    print()

    print("--- Python GC model ---")
    print("  1. Reference counting (primary)")
    print("     - Every object has a refcount")
    print("     - Freed immediately when refcount reaches 0")
    print("     - No pause -- deterministic lifetime")
    print()
    print("  2. Cyclic GC (supplementary)")
    print("     - Handles reference cycles (A->B->A)")
    print("     - Runs periodically based on allocation/deallocation thresholds")
    print("     - 3 generations: gen0 (young), gen1, gen2 (old)")
    print("     - Can cause small pauses")
    print()
    print("  vs Go GC:")
    print("     - Go: concurrent mark-and-sweep, sub-ms STW pauses")
    print("     - Python: refcount (instant) + cyclic GC (periodic)")
    print("     - Go: tunable via GOGC; Python: tunable via gc.set_threshold()")


if __name__ == "__main__":
    main()
